File: lib/src/text/text_normalizer.dart
/// Text normalization shared by the matching engine and dictionary storage.
///
/// Two distinct needs are served here:
///
///  * [TextNormalizer.normalizeKey] produces a canonical, length-*insensitive*
///    key used to compare terms (deduplicate dictionary entries, look entries
///    up). It may collapse whitespace and fold case.
///
///  * [TextNormalizer.normalizeToken] normalizes a single already-isolated
///    word *without* changing the relationship to source character offsets in
///    any way that matters for matching (it only lower-cases and unifies a few
///    punctuation variants, transforms that are 1:1 per code unit for the Latin
///    text we care about). The matcher tokenizes the original page text and
///    normalizes each token with this method, so reported match offsets still
///    line up with the PDF's per-character bounding boxes.
library;

/// Stateless helpers; grouped in a class purely for namespacing.
abstract final class TextNormalizer {
  /// Characters that are treated as equivalent to a straight apostrophe.
  static const _apostrophes = "'’ʼ‘`´";

  /// Characters treated as equivalent to an ASCII hyphen-minus.
  static const _hyphens = '-‐‑‒–—−';

  /// Normalize a single token (a "word"): lower-case and unify apostrophe and
  /// hyphen variants. Length is preserved for the scripts the app targets, so
  /// callers may keep using the token's original source offsets.
  static String normalizeToken(String token) {
    final buffer = StringBuffer();
    for (final rune in token.runes) {
      var ch = String.fromCharCode(rune);
      if (_apostrophes.contains(ch)) {
        ch = "'";
      } else if (_hyphens.contains(ch)) {
        ch = '-';
      }
      buffer.write(ch.toLowerCase());
    }
    return buffer.toString();
  }

  /// Normalize an arbitrary string into a canonical lookup/dedup key:
  /// trim, unify punctuation, lower-case, and collapse internal whitespace to
  /// single spaces.
  static String normalizeKey(String input) {
    final tokens = tokenize(input).map((t) => normalizeToken(t.text));
    return tokens.join(' ');
  }

  /// Split [input] into word tokens with their source offsets.
  ///
  /// A token is a maximal run of Unicode letters/digits, optionally joined by a
  /// single apostrophe or hyphen between two such runs (so `don't`,
  /// `well-known` and `café` are single tokens). Offsets are code-unit indices
  /// into [input], matching how Dart `String` and the PDF `charRects` list are
  /// indexed.
  static List<Token> tokenize(String input) {
    final result = <Token>[];
    for (final m in _tokenPattern.allMatches(input)) {
      result.add(Token(m.start, m.end, m.group(0)!));
    }
    return result;
  }

  /// The connector characters allowed inside a word, as a regex
  /// character-class body. The ASCII hyphen is escaped (`\-`) so it isn't
  /// interpreted as a range operator within the class.
  static const _wordConnectors = r"'’ʼ‘`´\-‐‑‒–—−";

  static final RegExp _tokenPattern = RegExp(
    '[\\p{L}\\p{N}]+(?:[$_wordConnectors][\\p{L}\\p{N}]+)*',
    unicode: true,
  );
}

/// A single word token and its half-open source range `[start, end)`.
class Token {
  const Token(this.start, this.end, this.text);

  final int start;
  final int end;
  final String text;

  @override
  String toString() => 'Token($start..$end "$text")';
}

File: ios-native/Eng/Text/TextNormalizer.swift
import Foundation

/// A single word token and its half-open source range `[start, end)` in **UTF-16
/// code-unit** offsets — the same unit PDFKit's `NSRange`/`characterBounds(at:)`
/// use, so a match range can be handed straight to `PDFPage`.
struct Token {
    let start: Int
    let end: Int
    let text: String
}

/// Text normalization shared by the matching engine and dictionary storage —
/// a direct port of the Flutter app's `TextNormalizer`.
///
/// Two distinct needs are served here:
///  * `normalizeKey` produces a canonical, length-*insensitive* key used to
///    compare/deduplicate terms (may collapse whitespace, fold case).
///  * `normalizeToken` normalizes a single isolated word **1:1 per code unit**
///    (only case-fold + unify a few punctuation variants), so match offsets
///    computed against a normalized token still index into the raw page text
///    and its per-character rectangles.
enum TextNormalizer {
    /// Characters treated as equivalent to a straight apostrophe.
    private static let apostrophes = Set("'’ʼ‘`´".unicodeScalars)
    /// Characters treated as equivalent to an ASCII hyphen-minus.
    private static let hyphens = Set("-‐‑‒–—−".unicodeScalars)

    /// Normalize a single token (a "word"): lower-case and unify apostrophe and
    /// hyphen variants. Length is preserved for the scripts the app targets, so
    /// callers may keep using the token's original source offsets.
    static func normalizeToken(_ token: String) -> String {
        var out = String.UnicodeScalarView()
        out.reserveCapacity(token.unicodeScalars.count)
        for scalar in token.unicodeScalars {
            if apostrophes.contains(scalar) {
                out.append("'")
            } else if hyphens.contains(scalar) {
                out.append("-")
            } else {
                // Lower-case per scalar (1:1 for the Latin text we target).
                out.append(contentsOf: String(scalar).lowercased().unicodeScalars)
            }
        }
        return String(out)
    }

    /// Normalize an arbitrary string into a canonical lookup/dedup key: unify
    /// punctuation, lower-case, and collapse internal whitespace to single spaces.
    static func normalizeKey(_ input: String) -> String {
        tokenize(input).map { normalizeToken($0.text) }.joined(separator: " ")
    }

    // The connector characters allowed inside a word, as a regex character-class
    // body. The ASCII hyphen is escaped so it isn't read as a range operator.
    private static let wordConnectors = #"'’ʼ‘`´\-‐‑‒–—−"#

    private static let tokenRegex: NSRegularExpression = {
        let pattern = "[\\p{L}\\p{N}]+(?:[\(wordConnectors)][\\p{L}\\p{N}]+)*"
        return try! NSRegularExpression(pattern: pattern)
    }()

    /// Split `input` into word tokens with their UTF-16 source offsets.
    ///
    /// A token is a maximal run of Unicode letters/digits, optionally joined by a
    /// single apostrophe or hyphen between two such runs (so `don't`,
    /// `well-known` and `café` are single tokens).
    static func tokenize(_ input: String) -> [Token] {
        let ns = input as NSString
        let full = NSRange(location: 0, length: ns.length)
        var result: [Token] = []
        tokenRegex.enumerateMatches(in: input, range: full) { match, _, _ in
            guard let r = match?.range, r.location != NSNotFound else { return }
            result.append(Token(start: r.location,
                                 end: r.location + r.length,
                                 text: ns.substring(with: r)))
        }
        return result
    }

    private static let edgePunctuation =
        try! NSRegularExpression(pattern: "^[^\\p{L}\\p{N}]+|[^\\p{L}\\p{N}]+$")

    /// Strip leading/trailing non-letter/digit characters while leaving inner
    /// word(s) and internal apostrophes/hyphens intact: `"oblate,"` -> `oblate`,
    /// `"(don't)."` -> `don't`.
    static func trimEdgePunctuation(_ input: String) -> String {
        let ns = input as NSString
        return edgePunctuation.stringByReplacingMatches(
            in: input, range: NSRange(location: 0, length: ns.length), withTemplate: "")
    }

    // PDFium delivers the hyphen it inserts at a line break as U+0002; a literal
    // soft hyphen (U+00AD) is treated the same. Stripping the marker plus hugging
    // whitespace/newline rejoins the word: "undis␂turbed" -> "undisturbed".
    // Embed the actual U+0002 / U+00AD scalars in the class (Swift string escapes,
    // one backslash) — ICU regex does NOT accept the "\u{XXXX}" brace escape, and a
    // "\\u{...}" regex escape here throws, crashing on the first selection.
    private static let pdfiumHyphen =
        try! NSRegularExpression(pattern: "[ \\t]*[\u{0002}\u{00AD}][ \\t]*(?:\\r?\\n)?[ \\t]*")
    // A word split by a *literal* visible hyphen at a line wrap.
    private static let wrapHyphen =
        try! NSRegularExpression(pattern: "(\\p{L})[-‐‑][ \\t]*(?:\\r?\\n)+[ \\t]*(\\p{Ll})")
    private static let lineBreaks =
        try! NSRegularExpression(pattern: "[ \\t]*(?:\\r?\\n)+[ \\t]*")

    /// Rewrite text selected across several lines into continuous text: rejoin a
    /// word hyphenated at a line wrap (letter before the hyphen, lower-case letter
    /// after — the signature of soft hyphenation), then collapse remaining line
    /// breaks to single spaces.
    static func joinWrappedLines(_ input: String) -> String {
        if input.isEmpty { return input }
        func replace(_ re: NSRegularExpression, _ s: String, _ tmpl: String) -> String {
            let ns = s as NSString
            return re.stringByReplacingMatches(
                in: s, range: NSRange(location: 0, length: ns.length), withTemplate: tmpl)
        }
        let joined = replace(pdfiumHyphen, input, "")
        let dehyphenated = replace(wrapHyphen, joined, "$1$2")
        return replace(lineBreaks, dehyphenated, " ")
    }
}
